pilot/pkg/proxy/envoy/discovery.go

```diff
@@ -432,23 +442,40 @@ func (ds *DiscoveryService) clearCache() {
 	clearCacheMutex.Lock()
 	defer clearCacheMutex.Unlock()
 
-	if time.Since(lastClearCache) < time.Duration(clearCacheTime)*time.Second {
-		if !clearCacheTimerSet {
-			clearCacheTimerSet = true
-			time.AfterFunc(time.Duration(clearCacheTime)*time.Second, func() {
-				clearCacheMutex.Lock()
-				clearCacheTimerSet = false
-				clearCacheMutex.Unlock()
-				ds.clearCache() // it's after time - so will clear the cache
-			})
-		}
+	// If last config change was > 1 second ago, push.
+	if time.Since(lastClearCacheEvent) > 1*time.Second {
+		lastClearCacheEvent = time.Now()
+		lastClearCache = time.Now()
+		log.Infof("Cleared discovery service cache after 1 sec of quiet")
+		V2ClearCache()
+		return
+	}
+
+	lastClearCacheEvent = time.Now()
+
+	// If last config change was < 1 second ago, but last push is > clearCacheTime ago -
+	// also push
+
+	if time.Since(lastClearCache) > time.Duration(clearCacheTime)*time.Second {
+		log.Infof("Cleared discovery service cache after %v", time.Since(lastClearCache))
+		lastClearCache = time.Now()
```

Comment:
This block seems unnecessary given that the previous if block would have triggered a push on the first event in a burst of config events

Reply:
It's for cases where changes are made continuously - for example a larger system where an endpoint or config changes every 1/2 second for a longer time. We'll still  do pushes every clearCacheTime. We can't do them every second because it would kill pilot (memory overhead is still high).